### PR TITLE
Updated "Made it work with the 2023-10 version of the API"

### PR DESCRIPTION
**Pull request by pevner-p2**
Please see https://github.com/ProdPerfect/monday/pull/116!

This is an updated version of that PR that better complies with the internal ProdPerfect linter, and also cleans up some of the issues noted in that PR.

Contributor's Note:   I've done my best just to address existing issues and stick only to changes requested in that PR.  I appreciate that this repo could use some love, but I'd like to get a release out to respond to the API update first.

### CODE REVIEW

File: Pipfile [hunk 1]
```diff
@@ -12,4 +12,4 @@ pytest = "*"
 [packages]
 
 [requires]
-python_version = "3.7"
+python_version = "3.8"
```
Comment (pevner-p2) [inline Pipfile:15]: Would like to confirm that this okay, as we currently say that we support all the way back to 3.6.

Comment (rhymiz) [reply Pipfile:15]: This is ok. Should also update the docs

Comment (pevner-p2) [reply Pipfile:15]: It breaks the test build, so it's going.

**Comment by pevner-p2** [comment]
@willnaoosmith @matiboy I did some updates addressing stated concerns, figured it was easier to have them as a fresh PR to the ProdPerfect repo than to send them to Will's repo and then back again.

File: monday/client.py [hunk 2]
```diff
@@ -7,23 +7,34 @@
...
 from .__version__ import __version__
-from .resources import ItemResource, UpdateResource, TagResource, BoardResource, UserResource, GroupResource, ComplexityResource, WorkspaceResource, NotificationResource
+from .resources import CustomResource, ItemResource, ColumnsResource, UpdateResource, TagResource, BoardResource, \
+    UserResource, GroupResource, ComplexityResource, WorkspaceResource, NotificationResource, MeResource
+
+_DEFAULT_HEADERS = {}
 
 
 class MondayClient:
-    def __init__(self, token):
```
Comment (chdastolfo) [inline monday/client.py:14]: Curious about this change - is this a preference thing? Monday does call them "tokens" rather than API keys, so token does reflect their internal language.

**Review by chdastolfo — APPROVED** [review]
Put the test build in the list of things that this repo needs to spruce things up that is oldddddd. Thanks so much for doing this! Changes look good, I'm going to run a few tests locally tomorrow but am happy to merge this in.

File: monday/client.py (hunk 2 above)
Comment (pevner-p2) [reply monday/client.py:14]: Darnit - we've been using both terms somewhat interchangeably in the code, and I thought I was helping to consolidate us.  However, I went in the wrong direction.  I'm going to make an update to fix that.

File: monday/graphqlclient/client.py [hunk 3]
```diff
@@ -1,41 +1,56 @@
...
+            headers.setdefault('content', 'multipart/form-data')
+                
             files = [
                 ('variables[file]', (variables['file'], open(variables['file'], 'rb')))
             ]
 
         try:
             response = requests.request("POST", self.endpoint, headers=headers, data=payload, files=files)
             response.raise_for_status()
+            self._catch_error(response)
```
Comment (rhymiz) [inline monday/graphqlclient/client.py:48]: this is not really "catching" errors - it's more validating and throwing an exception if there are errors, so we should probably call it `_throw_on_error(response)` or something

File: monday/graphqlclient/client.py [hunk 4]
```diff
@@ -1,41 +1,56 @@
...
             return response.json()
-        except requests.HTTPError as e:
+        except (requests.HTTPError, MondayQueryError) as e:
             print(e)
```
Comment (matiboy) [inline monday/graphqlclient/client.py:51]: It'd probably be nice to start having "proper" logging using the Python's standard library. Maybe a future feature. 
In any case, the print should go imho

File: monday/graphqlclient/client.py [hunk 5]
```diff
@@ -1,41 +1,56 @@
...
             raise e
+
+    def _catch_error(self, response):
+        if 'errors' in response.json():
+            raise MondayQueryError(response.json()['errors'][0]['message'])
```
Comment (matiboy) [inline monday/graphqlclient/client.py:56]: Can't help but feel wrong about reducing the errors to the first one only; Monday sends multiple errors when they can, would be nice if we can keep that somehow; maybe `MondayQueryError` could have an extra `original_errors` prop

File: monday/graphqlclient/client.py [hunk 6]
```diff
@@ -1,41 +1,56 @@
...
             response.raise_for_status()
+            self._catch_error(response)
             return response.json()
-        except requests.HTTPError as e:
+        except (requests.HTTPError, MondayQueryError) as e:
             print(e)
             raise e
+
+    def _catch_error(self, response):
+        if 'errors' in response.json():
```
Comment (matiboy) [inline monday/graphqlclient/client.py:55]: Risk of triggering a JSON parse exception which is not caught by the surrounding try catch (only HTTPError and MnodayError caught)

Comment (matiboy) [reply monday/graphqlclient/client.py:55]: Also we're parsing twice and given that Monday can return very large data, that's a bit of a waste of resources

**Review by matiboy — COMMENTED** [review]
Just a little question mark on the _catch_errors
but really happy that this library is being updated so quickly to update the API's breaking changes! 👍

File: monday/graphqlclient/client.py (hunk 3 above)
Comment (rhymiz) [reply monday/graphqlclient/client.py:48]: also, maybe?

```python
json_response = response.json()
self._raise_for_error(json_response)
return json_response
```

```python
def _raise_for_error(self, data)
    if "errors" in data:
        raise MondayQueryError(data["errors"])
```

File: monday/graphqlclient/client.py [hunk 7]
```diff
@@ -1,41 +1,56 @@
...
     def _send(self, query, variables):
         payload = {'query': query}
-        headers = {}
+        headers = self.headers.copy()
         files = None
-
+        
         if self.token is not None:
-            headers[self.headername] = '{}'.format(self.token)
+            headers[self.token_header_name] = '{}'.format(self.token)
```
[the remark this answers is not in the document]
Comment (pevner-p2) [reply monday/graphqlclient/client.py]: I absolutely hate it, and I have no idea why it was ever there.  I resisted messing with it to keep with the scope of the PR, but you motivated me to burn it.

File: monday/graphqlclient/client.py (hunk 4 above)
Comment (pevner-p2) [reply monday/graphqlclient/client.py]: Done, and agreed, some real logging in here would be great.

File: monday/graphqlclient/client.py (hunk 5 above)
Comment (pevner-p2) [reply monday/graphqlclient/client.py]: As you wish, so it is done.

File: monday/graphqlclient/client.py (hunk 6 above)
Comment (pevner-p2) [reply monday/graphqlclient/client.py]: Done and done.

File: monday/graphqlclient/client.py (hunk 3 above)
Comment (pevner-p2) [reply monday/graphqlclient/client.py]: `_throw_on_error` it is.

Comment (pevner-p2) [reply monday/graphqlclient/client.py]: Also, tell me how you feel about the `MondayQueryError` changes.  I picked a direction and went with it, but I could see more than a few other ways to do it.

File: monday/client.py [hunk 8]
```diff
@@ -7,23 +7,34 @@
...
-        self.tags = TagResource(token=token)
-        self.boards = BoardResource(token=token)
-        self.users = UserResource(token=token)
-        self.groups = GroupResource(token=token)
-        self.complexity = ComplexityResource(token=token)
-        self.workspaces = WorkspaceResource(token=token)
-        self.notifications = NotificationResource(token=token)
+
+        if not headers:
+            headers = _DEFAULT_HEADERS.copy()
```
Comment (matiboy) [inline monday/client.py:24]: Give that this is 2.0.0 and that the goal is to switch to new API, should default headers be `{"API-Version": "2023-10"}` from the get-go?

File: monday/client.py [hunk 9]
```diff
@@ -7,23 +7,34 @@
...
+        self.boards = BoardResource(token=token, headers=headers)
+        self.users = UserResource(token=token, headers=headers)
+        self.groups = GroupResource(token=token, headers=headers)
+        self.complexity = ComplexityResource(token=token, headers=headers)
+        self.workspaces = WorkspaceResource(token=token, headers=headers)
+        self.notifications = NotificationResource(token=token, headers=headers)
+        self.me = MeResource(token=token, headers=headers)
 
     def __str__(self):
         return f'MondayClient {__version__}'
```
Comment (matiboy) [inline monday/client.py:40]: Worth having the API Version as part of the string rep of the client maybe?

**Review by matiboy — COMMENTED** [review]
👍 Great!

File: monday/resources/columns.py [hunk 10]
```diff
@@ -0,0 +1,13 @@
+from monday.query_joins import update_multiple_column_values_query, create_column
+
+from monday.resources.base import BaseResource
+
+
+class ColumnsResource(BaseResource):
+    def create_column(self, board_id):
+        query = create_column(board_id)
```
Comment (chdastolfo) [inline monday/resources/columns.py:8]: `column_title` is also required for this method.

File: monday/resources/columns.py [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+        return self.client.execute(query)
+
+    def update_multiple_columns(self, item_id, board_id, column_values):
+        query = update_multiple_column_values_query(item_id, board_id, column_values)
```
Comment (chdastolfo) [inline monday/resources/columns.py:12]: Additionally, the `update_multiple_column_values_query` method has an optional parameter `create_labels_if_missing` that is missing here - it needs to be added as an optional argument to this signature or calls to this will fail.

File: monday/resources/columns.py [hunk 12]
```diff
@@ -0,0 +1,13 @@
+from monday.query_joins import update_multiple_column_values_query, create_column
```
Comment (chdastolfo) [inline monday/resources/columns.py:1]: This resource doesn't exist in the current version of the package, and after testing, neither method works. See comments below. Not sure if you want to fix or just not include @pevner-p2 @rhymiz?

**Comment by chdastolfo** [comment]
@pevner-p2 and @rhymiz a little late to start this discussion, but I wonder if there's an MVP in terms of what absolutely needs to be merged here to make this package work with the new API. I tested the added duplicate_board functionality and the Me resource and those seem to work ok, and I do like the idea of adding a resource to create custom queries, I know we talked about that a loooong time ago. Neither of the methods in the newly added Column resource are functional.

Obviously this PR is bigger than it strictly needs to be, which I realize we're all probably thinking about. I don't want to merge this with that Column resource until those methods are either fixed or removed. @pevner-p2 I still have my old Monday test account, so I'm happy to try to fix them if you're cool with me pushing to your branch. Thoughts?

File: monday/client.py (hunk 8 above)
Comment (pevner-p2) [reply monday/client.py:26]: That makes sense to me, done.

File: monday/client.py (hunk 9 above)
Comment (pevner-p2) [reply monday/client.py:42]: Probably, but I'm not sure if it should be the default API version or the one actually used by that specific client.  I think specific client makes more sense, but I could see an argument for the other way.  Either way, I'm skipping this for the moment, as it's not exactly essential.

File: monday/resources/columns.py (hunk 10 above)
Comment (pevner-p2) [reply monday/resources/columns.py]: Fixed.

File: monday/resources/columns.py (hunk 11 above)
Comment (pevner-p2) [reply monday/resources/columns.py]: Fixed.

**Review by matiboy — COMMENTED** [review]
Great 👍

File: monday/graphqlclient/client.py [hunk 13]
```diff
@@ -1,41 +1,56 @@
...
         self.token = None
-        self.headername = None
+        self.token_header_name = None
+        self.headers = {}
 
     def execute(self, query, variables=None):
         return self._send(query, variables)
 
-    def inject_token(self, token, headername='Authorization'):
+    def inject_token(self, token, token_header_name='Authorization'):
```
Comment (rhymiz) [inline monday/graphqlclient/client.py:18]: will the header name ever be named differently? if not, can probably scrap this bit

File: monday/resources/boards.py [hunk 14]
```diff
@@ -1,28 +1,22 @@
...
-from monday.resources.types import BoardKind, BoardState, BoardsOrderBy
+from monday.resources.types import BoardKind, BoardState, BoardsOrderBy, DuplicateType
 
 
 class BoardResource(BaseResource):
-    def __init__(self, token):
-        super().__init__(token)
-
-    def fetch_boards(self, limit: int = None, page: int = None, ids: List[int] = None, board_kind: BoardKind = None, state: BoardState = None, order_by: BoardsOrderBy = None):
+    def fetch_boards(self, limit: int = None, page: int = None, ids: List[int] = None, board_kind: BoardKind = None,
```
[the remark this answers is not in the document]
Comment (rhymiz) [reply monday/resources/boards.py:10]: I prefer to use `Union[TYPE, None]` but no strong feelings at this moment.

File: monday/client.py [hunk 15]
```diff
@@ -7,23 +7,36 @@
...
+    UserResource, GroupResource, ComplexityResource, WorkspaceResource, NotificationResource, MeResource
+
+_DEFAULT_HEADERS = {
+    "API-Version": "2023-10"
+}
 
 
 class MondayClient:
-    def __init__(self, token):
+    def __init__(self, token, headers=None):
```
Comment (rhymiz) [inline monday/client.py:19]: [not blocking] seems like we're not using typing consistently in this project. 
We should probably put up another PR at some point to address that.

File: monday/resources/boards.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (pevner-p2) [reply monday/resources/boards.py]: Yeah, as has been remarked elsewhere, we're probably going to need a PR dedicated to making typing usage correct and consistent.  I've fixed up this one, but I think there's definitely a lot to go with that.

File: monday/graphqlclient/client.py (hunk 13 above)
Comment (pevner-p2) [reply monday/graphqlclient/client.py]: Done.

**Comment by pevner-p2** [comment]
I think we're in a reasonable place to get this deployed, anything I'm missing @chdastolfo @rhymiz?

**Review by rhymiz — APPROVED** [review]
🥇 🥈 🥉
